File: src/Q41.cpp
// 对顶堆
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class MedianFinder {
private:
  priority_queue<int> queMin;
  // priority_queue<int, vector<int>, less<int>> queMin;
  priority_queue<int, vector<int>, greater<int>> queMax;

public:
  /** initialize your data structure here. */
  MedianFinder() {}

  void addNum(int num) {
    if (queMin.empty() || queMin.top() > num) {
      queMin.push(num);
      if (queMin.size() > queMax.size() + 1) {
        queMax.push(queMin.top());
        queMin.pop();
      }
    } else {
      queMax.push(num);
      if (queMax.size() > queMin.size()) {
        queMin.push(queMax.top());
        queMax.pop();
      }
    }
  }

  double findMedian() {
    if (queMin.size() > queMax.size()) {
      return double(queMin.top());
    } else {
      return double(queMin.top() + queMax.top()) / 2;
    }
  }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: src/Q41.rs
// 对顶堆法
use std::{cmp::Reverse, collections::BinaryHeap, iter::Rev};

struct MedianFinder {
    heap_max: BinaryHeap<Reverse<i32>>,
    heap_min: BinaryHeap<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MedianFinder {
    /** initialize your data structure here. */
    fn new() -> Self {
        Self {
            heap_max: BinaryHeap::new(),
            heap_min: BinaryHeap::new(),
        }
    }

    fn add_num(&mut self, num: i32) {
        if self.heap_min.is_empty() || *self.heap_min.peek().unwrap() > num {
            self.heap_min.push(num);
            if self.heap_min.len() > self.heap_max.len() + 1 {
                self.heap_max.push(Reverse(self.heap_min.pop().unwrap()));
            }
        } else {
            self.heap_max.push(Reverse(num));
            if self.heap_max.len() > self.heap_min.len() {
                self.heap_min.push(self.heap_max.pop().unwrap().0);
            }
        }
    }

    fn find_median(&self) -> f64 {
        if self.heap_min.len() > self.heap_max.len() {
            *self.heap_min.peek().unwrap() as f64
        } else {
            (*self.heap_min.peek().unwrap() + self.heap_max.peek().unwrap().0) as f64 / 2.0
        }
    }
}
